### PR TITLE
feature/octree_raycast

### PR DESCRIPTION
**Pull request by TheEmidee**
Created a set of classes USVORayCaster to allow to choose between different ray casts implementation at run time for the Theta* and Lazy Theta * algorithms to compute line of sight between 2 points.

There are ray casts which rely on the physics engine LineTrace and SphereTrace.
There's also a parametric algorithm which uses the properties of the octree to quickly compute if a ray intersects an occluded voxel or not

### CODE REVIEW

File: Source/SVONavigation/Private/Raycasters/SVORaycaster.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+{
+    UE_LOG( LogTemp, Warning, TEXT( "SubNode Address : %i - %i - %i" ), node_address.LayerIndex, node_address.NodeIndex, node_address.SubNodeIndex );
+    DebugInfos.TraversedLeafSubNodes.Emplace( node_address, is_occluded );
+}
+
+bool USVORayCaster::Trace( UObject * world_context, const FVector & from, const FVector & to, const FNavAgentProperties & nav_agent_properties ) const
+{
+    if ( UNavigationSystemV1 * navigation_system = UNavigationSystemV1::GetCurrent( world_context->GetWorld() ) )
+    {
+        if ( auto * navigation_data = navigation_system->GetNavDataForProps( nav_agent_properties ) )
```
Comment (Frickerson) [inline Source/SVONavigation/Private/Raycasters/SVORaycaster.cpp:66]: const auto *

File: Source/SVONavigation/Private/Raycasters/SVORaycasterTest.cpp [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+    Super::DestroyRenderState_Concurrent();
+}
+
+FPrimitiveSceneProxy * USVORayCasterRenderingComponent::CreateSceneProxy()
+{
+    FSVORayCasterSceneProxyData proxy_data;
+    proxy_data.GatherData( *GetRayCasterTest() );
+
+    if ( FSVORayCasterSceneProxy * new_scene_proxy = new FSVORayCasterSceneProxy( *this, proxy_data ) )
```
Comment (Frickerson) [inline Source/SVONavigation/Private/Raycasters/SVORaycasterTest.cpp:191]: if statement not necessary

File: Source/SVONavigation/Private/SVONavDataRenderingComponent.cpp [hunk 3]
```diff
@@ -50,15 +49,40 @@ FSVONavigationMeshSceneProxy::FSVONavigationMeshSceneProxy( const UPrimitiveComp
             Boxes.Emplace( navigation_bounds_data.GetData().GetNavigationBounds(), FColor::White );
         }
 
-        const auto try_add_voxel_to_boxes = [ this, debug_infos ]( const FVector & voxel_location, const float voxel_half_extent, const bool is_occluded ) {
+        const auto try_add_node_text_infos = [ this, &debug_infos, &navigation_bounds_data ]( const MortonCode node_morton_code, const LayerIndex node_layer_index, const FVector & node_position ) {
```
Comment (Frickerson) [inline Source/SVONavigation/Private/SVONavDataRenderingComponent.cpp:52]: navigation_bounds_data isn't used

File: Source/SVONavigation/Private/SVONavigationTypes.cpp [hunk 4]
```diff
@@ -14,6 +14,19 @@ void FSVOLeaves::Reset()
...
 
+FSVOVolumeNavigationDataDebugInfos::FSVOVolumeNavigationDataDebugInfos():
+    bDebugDrawBounds( false ),
+    bDebugDrawNodeAddress( false ),
+    bDebugDrawMortonCoords( false ),
+    bDebugDrawNodeLocation( false ),
+    bDebugDrawLayers( false ),
+    LayerIndexToDraw( 1 ),
+    bDebugDrawSubNodes( false ),
+    DebugDrawOccludedVoxels( true ),
```
Comment (Frickerson) [inline Source/SVONavigation/Private/SVONavigationTypes.cpp:25]: prefix last two with 'b'

File: Source/SVONavigation/Private/SVOVolumeNavigationData.cpp [hunk 5]
```diff
@@ -45,28 +45,52 @@ FVector FSVOVolumeNavigationData::GetNodePositionFromAddress( const FSVONodeAddr
...
     const auto morton_coords = FSVOHelpers::GetVectorFromMortonCode( address.NodeIndex );
     const auto & navigation_bounds = SVOData.GetNavigationBounds();
 
-    auto position = navigation_bounds.GetCenter() - navigation_bounds.GetExtent() + morton_coords * voxel_size + voxel_half_size;
+    const auto navigation_bounds_center = navigation_bounds.GetCenter();
+    const auto navigation_bounds_extent = navigation_bounds.GetExtent();
+    auto position = navigation_bounds_center - navigation_bounds_extent + morton_coords * voxel_size + voxel_half_size;
 
-    if ( address.LayerIndex == 0 && node.FirstChild.IsValid() )
+   /* if ( address.LayerIndex == 0 && address.NodeIndex < static_cast< uint_fast32_t >( layer.GetNodeCount() ) )
```
Comment (Frickerson) [inline Source/SVONavigation/Private/SVOVolumeNavigationData.cpp:58]: remove commented code

File: Source/SVONavigation/Public/Raycasters/SVORaycaster.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+    FSVONodeAddress NodeAddress;
+    bool bIsOccluded;
+};
+
+struct FSVORayCasterDebugInfos
+{
+    FVector RayCastStartLocation;
+    FVector RayCastEndLocation;
+    bool Result;
```
Comment (Frickerson) [inline Source/SVONavigation/Public/Raycasters/SVORaycaster.h:26]: prefix with 'b'

File: Source/SVONavigation/Private/Raycasters/SVORaycaster.cpp (hunk 1 above)
Comment (TheEmidee) [reply Source/SVONavigation/Private/Raycasters/SVORaycaster.cpp:66]: has been removed in later PRs

File: Source/SVONavigation/Private/Raycasters/SVORaycasterTest.cpp (hunk 2 above)
Comment (TheEmidee) [reply Source/SVONavigation/Private/Raycasters/SVORaycasterTest.cpp:191]: fixed in last PR

File: Source/SVONavigation/Private/SVONavDataRenderingComponent.cpp (hunk 3 above)
Comment (TheEmidee) [reply Source/SVONavigation/Private/SVONavDataRenderingComponent.cpp:52]: removed in latest PRs

File: Source/SVONavigation/Private/SVONavigationTypes.cpp (hunk 4 above)
Comment (TheEmidee) [reply Source/SVONavigation/Private/SVONavigationTypes.cpp:25]: Fixed in last PR

File: Source/SVONavigation/Private/SVOVolumeNavigationData.cpp (hunk 5 above)
Comment (TheEmidee) [reply Source/SVONavigation/Private/SVOVolumeNavigationData.cpp:58]: Does not exist anymore in latests PRs

File: Source/SVONavigation/Public/Raycasters/SVORaycaster.h (hunk 6 above)
Comment (TheEmidee) [reply Source/SVONavigation/Public/Raycasters/SVORaycaster.h:26]: Does not exist in latest PRs

**Pull request merged**